### PR TITLE
new endpoint for grouped stats

### PR DESCRIPTION
**Pull request by shahronak47**
Adding a new endpoint `/grouped-stats` which calls `wbpip:::gd_compute_pip_stats` . If you have a better name for this endpoint please let me know.

### CODE REVIEW

**Review by randrescastaneda — CHANGES_REQUESTED** [review]
Hi @shahronak47 , 

I could not make it work (see bellow). Can you please provide me with an example of how it works? Also, could you please add a few unit tests? 

```
// http://127.0.0.1:8080/api/v1/grouped-stats?welfare=0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1&weight=0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1

{
  "error": "500 - Internal server error",
  "message": "Error in strsplit(params$population, \",\"): non-character argument\n"
}
```

Thanks.

File: inst/plumber/v1/endpoints.R
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R]: They are potentially many things that can go wrong here. Passing too many values for welfare or population, passing different number of value for welfare and population, etc.

It would be better to have specific function to parse these parameters, so we can unit test them for edge cases and ensure they work as expected.

Comment (randrescastaneda) [reply inst/plumber/v1/endpoints.R]: I Think that passing too many values is not a problem. Yet, I think a limit (say 100obs) suffices for the sake of the completeness of the functions. However, the check that both, welfare and population vectors should be the same length is necessary (though I think that it is implemented already in `wbpip`).

Comment (tonyfujs) [reply inst/plumber/v1/endpoints.R]: Thanks @randrescastaneda We will still need some higher level validation so the API can return an informative to the end-users. Otherwise, they will just get uninformative 500 errors.

**Comment by tonyfujs** [comment]
> Hi @shahronak47 ,
> 
> I could not make it work (see bellow). Can you please provide me with an example of how it works? Also, could you please add a few unit tests?
> 
> ```
> // http://127.0.0.1:8080/api/v1/grouped-stats?welfare=0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1&weight=0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1
> 
> {
>   "error": "500 - Internal server error",
>   "message": "Error in strsplit(params$population, \",\"): non-character argument\n"
> }
> ```
> 
> Thanks.

@randrescastaneda You are getting this error because the you are passing a `weight` parameter in your query, when the endpoint is expecting a `population` parameter. There are also other parameters that are required for the underlying function to work:
image

A few thoughts about this:
- `requested_mean` should probably be named `mean`
- We should probably not expect end-users to specify their own ppp, so the `default_ppp` parameter should probably be removed and defined in the anonymous functions that underpins the endpoint.

**Comment by randrescastaneda** [comment]
Hi @shahronak47, I create this substask with instructions for this PR [CU-86872qgw9 Basic stats](https://app.clickup.com/t/86872qgw9)

Also, make sure to follow the directions of this [other task](https://app.clickup.com/t/86872qgna), regarding the development of these endpoints. 

Thanks.

**Comment by shahronak47** [comment]
> * `requested_mean` should probably be named `mean`

Where do we want to do this change? In the API or the underlying function `wbpip:::gd_compute_pip_stats`. Right now, I have kept all the arguments in sync with the function call. 

My personal opinion is the arguments to API and the function arguments should be in-sync so we should do this change in the function and then API will follow suit. Let me know what you think @randrescastaneda @tonyfujs

**Comment by tonyfujs** [comment]
> > * `requested_mean` should probably be named `mean`
> 
> Where do we want to do this change? In the API or the underlying function `wbpip:::gd_compute_pip_stats`. Right now, I have kept all the arguments in sync with the function call.
> 
> My personal opinion is the arguments to API and the function arguments should be in-sync so we should do this change in the function and then API will follow suit. Let me know what you think @randrescastaneda @tonyfujs

Hi @shahronak47. Good point, but in that case I don't think we should rename the function's argument as we have two various types of `mean` arguments in PIP. Let me check and get back to you.

File: inst/plumber/v1/endpoints.R
Comment (shahronak47) [reply inst/plumber/v1/endpoints.R]: I agree @tonyfujs , I have added a function to check inputs and return an informative error. Please have another look.

**Comment by shahronak47** [comment]
>We should probably not expect end-users to specify their own ppp, so the default_ppp parameter should probably be removed and defined in the anonymous functions that underpins the endpoint.

The underlying function has default value for ppp (`default_ppp`) as 1, the API takes the default value if nothing is provided.

**Comment by shahronak47** [comment]
@randrescastaneda I am using these values to test the API. 

Welfare - "0.0002,0.0006,0.0011,0.0021,0.0031,0.0048,0.0066,0.0095,0.0128,0.0177,0.0229,0.0355,0.0513,0.0689,0.0882" (Enter without quotes in Swagger UI)
Population -  "0.001,0.003,0.005,0.009,0.013,0.019,0.025,0.034,0.044,0.0581,0.0721,0.1041,0.1411,0.1792,0.2182" Enter without quotes in Swagger UI)
 requested_mean = 2.911786
povline = 1.9

**Comment by randrescastaneda** [comment]
> > We should probably not expect end-users to specify their own ppp, so the default_ppp parameter should probably be removed and defined in the anonymous functions that underpins the endpoint.
> 
> The underlying function has default value for ppp (`default_ppp`) as 1, the API takes the default value if nothing is provided.

Hi @shahronak47. Thank you, I understand now the logic of your decision. The bottom line is not to worry about it. You can remove `default_ppp` from the parameters of this new endpoint or leave the default as 1.

The reason is the following. There are two ppp parameters, `ppp` and `default_ppp`. `ppp` is only relevant when the `requested_mean` is in Local Currency Unit (i.e., the currency of the currency). If the `requested_mean` is in PPP values already, both `ppp` and `default_ppp` play no role in the calculations. `default_ppp` is only used when the user wants to "play" with the value of the PPP if it were different from they one in which the `requested_mean` already is. It works like an adjustment factor of the `requested_mean`. So, either remove it or leave it with 1 as the default. Leaving it with the default makes more sense, in case we want to give that option to experienced users, but let's not advertize it to much to avoid confusion. 

@tonyfujs , what do you think?

**Comment by randrescastaneda** [comment]
> The reason is the following. There are two ppp parameters, `ppp` and `default_ppp`. `ppp` is only relevant when the `requested_mean` is in Local Currency Unit (i.e., the currency of the currency). If the `requested_mean` is in PPP values already, both `ppp` and `default_ppp` play no role in the calculations. `default_ppp` is only used when the user wants to "play" with the value of the PPP if it were different from they one in which the `requested_mean` already is. It works like an adjustment factor of the `requested_mean`. So, either remove it or leave it with 1 as the default. Leaving it with the default makes more sense, in case we want to give that option to experienced users, but let's not advertize it to much to avoid confusion.
> 
> @tonyfujs , what do you think?

@tonyfujs, I defer to you this PR. I think it is ready to merge, but I leave it up to you. Also, given my explanation above, do you agree with removing `default_ppp` or should we leave it?

**Comment by shahronak47** [comment]
Also added the end point `regression-params` for this task https://app.clickup.com/t/86872qh1m . However, it uses `pipster` package and requires this PR to be approved https://github.com/PIP-Technical-Team/wbpip/pull/236

**Comment by shahronak47** [comment]
As discussed, I have copied the functions from `pipster` and `wbpip` which are not yet available. The endpoints are now ready to test.

**Comment by randrescastaneda** [comment]
> 2.911786

This worked really well. 

`http://127.0.0.1:8080/api/v1/grouped-stats?welfare=0.0002,0.0006,0.0011,0.0021,0.0031,0.0048,0.0066,0.0095,0.0128,0.0177,0.0229,0.0355,0.0513,0.0689,0.0882&population=0.001,0.003,0.005,0.009,0.013,0.019,0.025,0.034,0.044,0.0581,0.0721,0.1041,0.1411,0.1792,0.2182&requested_mean=2.911786&povline=1.9` 
image

**Review by randrescastaneda — CHANGES_REQUESTED** [review]
Hi @shahronak47 , 

I checked the endpoints and they are looking good. However, there are a few things to do. 

### general note
In general, all the endpoints should be serialized in the same way that the `pip`  endpoint works. That is, the user should be able to select between Json, Rds, or csv formats. Also, in order to avoid problems, make sure the output of each endpoint is in table format, not list. Otherwise, the csv format breaks

### specific changes
* **grouped-stats**: convert to table format and make sure the `format` argument is enabled. 
* **regression**: add defaults to the parameters and make sure the `povline` argument is also enabled. Also, suppressed the `p0` parameter. It does not do anything meaningful. 
* **lorenz**: I apologize for giving you misleading information, but the  correct function is not `wbpip:::sd_create_synth_vector`. You better use (copy and paste) all the functions in the [pipgd_lorenz.R](https://github.com/PIP-Technical-Team/pipster/blob/DEV/R/pipgd_lorenz.R) of the DEV branch. You may endup using function `pipgd_lorenz_curve()`. 
 

Please, let me know if you have any questions.

Best,

**Comment by shahronak47** [comment]
Hi @randrescastaneda , I have done the changes as per our last discussion : 

1. Include a column called `lorenz` to define two rows as lq and lb in `regression-params` endpoint
2. Change the column name to `welfare` and `weight` instead of `output` and `points` in `lorenz-curve` endpoint.

**Review by randrescastaneda — APPROVED** [review]

File: inst/plumber/v1/endpoints.R
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R]: You are already working with `args` as far as I can tell. Can this comment be removed?

File: inst/plumber/v1/endpoints.R
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R]: Has this been resolved?
- If not let's resolve it before merging
- If yes, can we remove this comment?

File: R/utils-plumber.R
Comment (tonyfujs) [inline R/utils-plumber.R]: I don't think `dplyr` is being used anywhere else in this package (it is not in the DESCRIPTION file). If there is no other use case, I don't think this is a sufficient use case to import such a large dependency as `dplyr`

File: inst/plumber/v1/endpoints.R
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R]: Same as above

File: inst/plumber/v1/endpoints.R
Comment (tonyfujs) [inline inst/plumber/v1/endpoints.R]: I think this logic could be simpler if the serializer is assigned at the beginning, no? Similar to what is done for the `pip-grp` endpoint for instance

**Review by tonyfujs — CHANGES_REQUESTED** [review]
Thanks @shahronak47 I made a few comments in line. In addition:
- I am getting the following error when I run `devtools::docment()`
image
- Please make sure to run `devtools::check()` I haven't ran it, but I'm pretty sure it does not pass currently because of the non-declared `dplyr` dependency.

Other than that it looks great! Let me know if any of my comment is not clear. Thanks!

File: tests/testthat/test-utils-plumber.R
Comment (tonyfujs) [inline tests/testthat/test-utils-plumber.R:483]: I added this unit test. Currently it does not pass, but I think this is a potentially valid edge case.

File: R/utils.R
Comment (randrescastaneda) [inline R/utils.R:411]: This should not have `max`. or something super big, like `1e10`.

File: R/utils.R
Comment (randrescastaneda) [inline R/utils.R:417]: Same here> This should not have `max`. or something super big, like `1e10`.

File: R/utils.R
Comment (randrescastaneda) [inline R/utils.R]: The min can't be zero. Also, I don't think this is parameter that should be included in the endpoint. This is an internal parameter in `pipster`. Does not have much functionality here. However, if it is already incorporated, do not take it away. Just make sure the min is `0.01` (1%), and the max value times the mean Can't be higher than the max value of the allowed poverty line, so this is trick. I would add a max of `5` and that's it.

File: vignettes/new-endpoints.Rmd
Comment (randrescastaneda) [inline vignettes/new-endpoints.Rmd:1]: This is fantastic. Well done!

**Review by randrescastaneda — CHANGES_REQUESTED** [review]
This is really good, @shahronak47 . THank you. I added a few small comments. Yet, we still need @tonyfujs 's stamp for approval. 
Thanks. Best,

**Comment by tonyfujs** [comment]
Hey @shahronak47  I am trying to run the most recent branch locally, and I'm getting errors. This is what I tried: http://127.0.0.1:8080/api/v1/grouped-stats?welfare=0.0002,0.0006,0.0011,0.0021,0.0031,0.0048,0.0066,0.0095,0.0128,0.0177,0.0229,0.0355,0.0513,0.0689,0.0882&population=0.001,0.003,0.005,0.009,0.013,0.019,0.025,0.034,0.044,0.0581,0.0721,0.1041,0.1411,0.1792,0.2182&requested_mean=2.911786&povline=1.9

image

**Comment by shahronak47** [comment]
Hi @tonyfujs , after discussion with Andres we switched the argument from `welfare` to `cum_welfare` and `population` to `cum_population` . Sorry for the confusion.  

Try using this  : 

http://127.0.0.1:8080/api/v1/grouped-stats?cum_welfare=0.0002,0.0006,0.0011,0.0021,0.0031,0.0048,0.0066,0.0095,0.0128,0.0177,0.0229,0.0355,0.0513,0.0689,0.0882&cum_population=0.001,0.003,0.005,0.009,0.013,0.019,0.025,0.034,0.044,0.0581,0.0721,0.1041,0.1411,0.1792,0.2182&requested_mean=2.911786&povline=1.9

**Comment by tonyfujs** [comment]
> grouped-stats?cum_welfare=0.0002,0.0006,0.0011,0.0021,0.0031,0.0048,0.0066,0.0095,0.0128,0.0177,0.0229,0.0355,0.0513,0.0689,0.0882&cum_population=0.001,0.003,0.005,0.009,0.013,0.019,0.025,0.034,0.044,0.0581,0.0721,0.1041,0.1411,0.1792,0.2182&requested_mean=2.911786&povline=1.9

Ok. Thanks @shahronak47, and sorry for not responding earlier. It working fine using those parameters. What I would suggest before we merge is to add documentation to the Swagger file for this new point, with a full example so end-users have a concrete example of how to use it. The specs for the Swagger documentation can be edited here: `./inst/plumber/v1/openapi.yaml`

**Comment by tonyfujs** [comment]
Also, @randrescastaneda  @shahronak47 can you both confirm that both `devtools::test()` and `devtools::check()` pass for you before merging. Some unit tests do not pass for me, but I'm quite sure this has to do with some bad configuration on my laptop, and not with the code itself. Thanks!

**Comment by shahronak47** [comment]
Hi @tonyfujs , I have updated the documentation in `openapi.yaml` file and I confirm that `devtools::test()` and `devtools::check()` passes for me with some minor notes. 

image

The first one we decided to keep it as it is https://github.com/PIP-Technical-Team/pipapi/issues/402 

cc : @randrescastaneda

**Comment by tonyfujs** [comment]
Great. Thanks @shahronak47 ! All good from my side then.

**Comment by shahronak47** [comment]
Thanks @tonyfujs , should we merge this to `DEV` then @randrescastaneda ?

**Review by randrescastaneda — APPROVED** [review]

**Review by randrescastaneda — APPROVED** [review]

**Pull request merged**
